src/isomorphic/ReactDebugTool.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright 2016-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactDebugTool
+ */
+
+'use strict';
+
+var ExecutionEnvironment = require('ExecutionEnvironment');
+var ReactPerfDevtool = require('ReactPerfDevtool');
+
+var warning = require('warning');
+
+var eventHandlers = [];
+var handlerDoesThrowForEvent = {};
+
+function emitEvent(handlerFunctionName, arg1, arg2, arg3, arg4, arg5) {
+  if (__DEV__) {
+    eventHandlers.forEach(function(handler) {
+      try {
+        if (handler[handlerFunctionName]) {
+          handler[handlerFunctionName](arg1, arg2, arg3, arg4, arg5);
+        }
+      } catch (e) {
+        warning(
+          !handlerDoesThrowForEvent[handlerFunctionName],
+          'exception thrown by devtool while handling %s: %s',
+          handlerFunctionName,
+          e.message
+        );
+        handlerDoesThrowForEvent[handlerFunctionName] = true;
+      }
+    });
+  }
+}
+
+var ReactDebugTool = {
+  addDevtool(devtool) {
+    eventHandlers.push(devtool);
+  },
+  removeDevtool(devtool) {
+    for (var i = 0; i < eventHandlers.length; i++) {
+      if (eventHandlers[i] === devtool) {
+        eventHandlers.splice(i, 1);
+        i--;
+      }
+    }
+  },
+  onBeginFlushUpdates() {
+    emitEvent('onBeginFlushUpdates');
+  },
+  onStartTimer(internalInstance, timerType) {
+    emitEvent('onStartTimer', internalInstance, timerType);
+  },
+  onEndFlushUpdates() {
+    emitEvent('onEndFlushUpdates');
+  },
+  onStopTimer(internalInstance, timerType) {
```

Comment:
Ah, I hadn't realised that we were completely removing nested calls. 

There is a lot of value in representing a heirarchy in profiling output. For example if you take a look at the image [here](https://gist.github.com/joshduck/0c35f70cdd3abf88770c) you immediately get a sense of which parts of the application need improvement. If you rely on aggregation then you lose that big picture view. This is the biggest problem with the current ReactDefaultPerf implementation. 

Would the work stack be completely divorced from the parent hierarchy? If it is still somewhat recognizable then there would be value in surfacing that information. Chrome profiler flame charts are still immensely valuable even though they can't follow asynchronous boundaries.